nav2_collision_monitor/src/polygon_base.cpp

```diff
@@ -0,0 +1,175 @@
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_base.hpp"
+#include "nav2_collision_monitor/kinematics.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+PolygonBase::PolygonBase()
+: polygon_type_(POLYGON_BASE), polygon_name_(""), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(0.0)
+{
+}
+
+PolygonBase::PolygonBase(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string polygon_name,
+  const double simulation_time_step)
+: node_(node), polygon_type_(POLYGON_BASE), polygon_name_(polygon_name), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(simulation_time_step)
+{
+  auto node_sptr = node_.lock();
+  if (node_sptr) {
+    RCLCPP_INFO(node_sptr->get_logger(), "Creating PolygonBase");
+  }
+}
+
+PolygonBase::~PolygonBase()
+{
+  auto node = node_.lock();
+  if (node) {
+    RCLCPP_INFO(node->get_logger(), "Destroying PolygonBase");
+  }
+}
+
+bool PolygonBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    // Get action type
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".action_type", rclcpp::ParameterValue("stop"));  // Stop by default
+    const std::string at_str =
+      node->get_parameter(polygon_name_ + ".action_type").as_string();
+    if (at_str == "stop") {
+      action_type_ = STOP;
+    } else if (at_str == "slowdown") {
+      action_type_ = SLOWDOWN;
+    } else {  // at_str == "approach"
+      action_type_ = APPROACH;
+    }
+
+    if (action_type_ == STOP) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+    } else if (action_type_ == SLOWDOWN) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".slowdown", rclcpp::ParameterValue(0.5));
+      slowdown_ = node->get_parameter(polygon_name_ + ".slowdown").as_double();
+    } else {  // action_type_ == APPROACH
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".time_before_collision",
+        rclcpp::ParameterValue(5.0));
+      time_before_collision_ =
+        node->get_parameter(polygon_name_ + ".time_before_collision").as_double();
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic polygon parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+PolygonType PolygonBase::getPolygonType()
```

Comment:
Are all these get/sets used? If any are not, remove them. This seems like alot of get/sets given that most of these variables are set in parameters that are found in this class itself.

Reply:
Optimized for polygons and sources